src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,160 @@
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState).ConfigureAwait(false);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+
+            return entry;
+        }
+
+        public static string FormatAuthorizeEndpoint(string host, string tenant)
+        {
+            return $"https://{host}/{tenant}/oauth2/authorize";
+        }
+
+        // No return value. Modifies InstanceCache directly.
+        private static async Task DiscoverAsync(string host, bool validateAuthority, CallState callState)
+        {
+            string tentativeAuthorizeEndpoint = FormatAuthorizeEndpoint(host, "irrelevant");
+            string instanceDiscoveryHost = WhitelistedAuthorities.Contains(host) ? host : DefaultTrustedAuthority;
+            string instanceDiscoveryEndpoint =
+                $"https://{instanceDiscoveryHost}/common/discovery/instance?api-version=1.1&authorization_endpoint={tentativeAuthorizeEndpoint}";
+            var client = new AdalHttpClient(instanceDiscoveryEndpoint, callState);
+            InstanceDiscoveryResponse discoveryResponse = null;
+            try
+            {
+                discoveryResponse = await client.GetResponseAsync<InstanceDiscoveryResponse>().ConfigureAwait(false);
+                if (validateAuthority && discoveryResponse.TenantDiscoveryEndpoint == null)
+                {
+                    // hard stop here
+                    throw new AdalException(AdalError.AuthorityNotInValidList);
+                }
+            }
+            catch (AdalServiceException ex)
+            {
+                // The pre-existing implementation (https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-e4febd8f40f03e71bcae0f990f9690eaL99)
+                // has been coded in this way: it catches the AdalServiceException and then translate it into 2 validation-relevant exceptions.
+                // So the following implementation absorbs these specific exceptions when the validateAuthority flag is false.
+                // All other unexpected exceptions will still bubble up, as always.
+                if (validateAuthority)
+                {
+                    // hard stop here
+                    throw new AdalException(
+                        (ex.ErrorCode == "invalid_instance")
```

Comment:
There should be caching of invalid_instance responses as well, we don't want to have to have an ADAL client purposefully hitting an invalid instance to always hit the discovery endpoint. This can be addressed in a follow up PR.

Reply:
Did we open an issue?